enhancements/insights/insights-operator-runtime.md

```diff
@@ -0,0 +1,532 @@
+
+### Non-Goals
+
+* It is not a goal of this proposal to scan container for security purposes
+* It is not a goal of this proposal to get runtime information specific to a given customer or application
+* It is not a goal of this proposal to capture personal sensitive information
+
+## Proposal
+
+This enhancement proposal introduces additional data to identify the technology stack of running containers.
+
+The existing `workloadContainerShape` type that contains workload information about a container will be enhanced with a new `workloadRuntimeInfoContainer` field:
+
+```go
+// workloadContainerShape describes the shape of a container which includes
+// a subset of the data in the container.
+// TODO: this may desirable to make more precise with a whole container hash
+//
+//  that includes more of the workload, but that would only be necessary if
+//  it assisted reconstruction of type of workloads.
+type workloadContainerShape struct {
+   // ImageID is the content addressable hash of the image as observed from
+   // the status or the spec tag.
+   ImageID string `json:"imageID"`
+   // FirstCommand is a hash of the first value in the command array, if
+   // any was set.
+   FirstCommand string `json:"firstCommand,omitempty"`
+   // FirstArg is a hash of the first value in the arguments array, if any
+   // was set.
+   FirstArg string `json:"firstArg,omitempty"`
+   // Runtime info on the container
+   RuntimeInfo workloadRuntimeInfoContainer `json:"runtimeInfo,omitempty"`
+}
+```
+
+This new `workloadRuntimeInfoContainer` is defined as:
+
+```go
+type workloadRuntimeInfoContainer struct {
+	// Hash of the identifier of the Operating System (based on /etc/os-release ID)
+	Os string `json:"os,omitempty"`
+	// Hash of the version identifier of the Operating System (based on /etc/os-release VERSION_ID)
+	OsVersion string `json:"osVersion,omitempty"`
+	// Identifier of the kind of runtime
+	Kind string `json:"kind,omitempty"`
+	// Version of the kind of runtime
+	KindVersion string `json:"kindVersion,omitempty"`
+	// Entity that provides the runtime-kind implementation
+	KindImplementer string `json:"kindImplementer,omitempty"`
+	// Runtimes components
+	Runtimes []RuntimeComponent `json:"runtimes,omitempty"`
+}
+
+type RuntimeComponent struct {
+	// Name of a runtime used to run the application in the container
+	Name string `json:"name,omitempty"`
+	// The version of this runtime
+	Version string `json:"version,omitempty"`
+}
+```
+
+These new data elements are all optional (as they may not be relevant depending on the type of workloads) and are hashed (and truncated) to preserve anonymity (the hashing algorithm is [the same one that is currently used by the gatherer](https://github.com/openshift/insights-operator/blob/30b161f98eb185bb357ac84f8aca58e369d34bcd/pkg/gatherers/workloads/workloads_info.go#L336)).
```

Comment:
I'm not clear on how truncated SHA-256 hashing of these strings is compatible with user stories like:

> As an OpenShift product marketing manager, I want to be able to produce content that shows users use OpenShift as a  platform for all types of workloads.

Is the idea that you can show a large range of OS hashes, etc., and handwave about that meaning there must be some meaningful diversity in inputs, without being able to say exactly which OS are being used?  Or will there be guessing about particular OS strings feeding a hash->original lookup for OS that the manager is interested in, in which case not all that much anonymity is being granted (unless you're using an OS string that the manager hasn't thought to look for).

Reply:
I see you do have:

> ...it is possible to create a dictionary of the hashes and retrieve the unobfuscated values...

below, but I think that makes the "preserve anonymity" point here less clear.  Can you add more detail about the kind of anonymity you're preserving?  Is it just "original strings that Red Hat hasn't thought to hash for the hash->original lookup table yet"?

Reply:
I amended the RFE and updated the wording for clarification:

> These new data elements are all optional (as they may not be relevant depending on the type of workloads) and are hashed (and truncated) to preserve shoulder-surfing anonymity (the hashing algorithm is [the same one that is currently used by the gatherer](https://github.com/openshift/insights-operator/blob/30b161f98eb185bb357ac84f8aca58e369d34bcd/pkg/gatherers/workloads/workloads_info.go#L336)).
> Someone with access to the date and who knows what they are looking for ("How many containers are running Go 1.21.3?") will be able to unobscure the data.